brille/design/Proposed_integrated_syntax_Brill_SpinW_Horace.md

```diff
@@ -0,0 +1,267 @@
+
+## Table of Contents
+
+[Introduction](#introduction)
+
+[Workflow](#workflow)
+
+[Extracting the data](#extracting-the-data)
+
+[Modelling the data](#modelling-the-data)
+
+[Fitting the data with resolution convolution](#fitting-the-data-with-resolution-convolution)
+
+[Implications for implementations](#implications-for-implementations)
+
+## Introduction
+
+[Brille](https://github.com/brille/brille) is a C++ library with Python wrappings for symmetry operations and linear interpolation within an irreducible part of the first Brillouin Zone.
+It can be used to construct the first Brillouin zone, find the high symmetry points and the irreducible polyhedron of any spacegroup.
+In addition, it can be used to linearly interpolate any quantity (vector or scalar) which obeys the symmetries of the input spacegroup.
+It is examples of this interpolating functionality that we shall explore in this document, specifically focusing on applications to magnetic time-of-flight neutron spectroscopy data.
+
+This document will include both working (mostly Matlab) and proposed (mostly Python) code examples.
+Proposed code examples will be marked by black lines on either side of the code block.
+The code examples will use a combination of Brille, SpinW, Horace and MSlice.
+
+## Workflow
+
+The workflow is:
+
+1. [Horace](https://github.com/pace-neutrons/Horace/) or *MSlice* is first used to extract (a subset of) the data to be modelled or fitted.
+2. SpinW or Brille is then used to simulate the extracted data with some set of model parameters. 
+The returned results should be of the same class as the data such that they may be easily plotted by Horace or MSlice.
+3. If the model has parameters (e.g. SpinW), it should be possible to fit those parameters to the measured data.
+4. For both fitting and simulation of a model, it should be possible to incorporate the instrumental resolution in the calculcation.
+
+## Extracting the data
+
+For single crystal data, this is currently possible in Matlab using Horace:
+
+```matlab
+proj = projaxes([1 0 0], [0 1 0]);
+cut1 = cut_sqw(sqw_file, proj, [hmin hstep hmax], [kmin kstep kmax], [lmin lmax], [emin estep emax]);
+```
+
+The above example assumes that data processing to construct an `.sqw` file combining multiple measurements of a crystal at different orientations is done.
+Triplets of numbers indicate that the data is to be rebinned along that axis (defined by the projection `proj`), whilst doublets indicates the data should be integrated between those ranges.
+Thus the example generates a 3D `sqw` object `cut1` from the 4D data in the `sqw_file`.
+
+The proposed syntax for Python is:
+
+<table>
+<td><img src="https://placehold.it/5x100/000000/000000?text=+"></td>
+<td>
+<pre lang="python">
+from pyHorace import horace as h
+proj = {'u':[1,0,0], 'v':[0,1,0], 'type':'rrr'}
+cut1 = h.cut_sqw(sqw_file, proj, [hmin, hstep, hmax], [kmin, kstep, kmax], 
+                                 [lmin, lmax], [emin, estep, emax])
```

Comment:
This is, effectively, 13 positional arguments and seems error-prone. Is it possible for some of the groups to be named optional arguments with useful defaults?

Also, pedantically, input groups like these are generally tuples rather than lists. So combining those we get a function signature like:
```python
def cut_sqw(sqw_file: Path, proj: Dict[str, Any],
            h_range: Optional[Tuple] = None, k_range: Optional[Tuple] = None,
            l_limits: Optional[Tuple] = None, e_range: Optional[Tuple] = None) -> WhateverTypeACutHas
```
and a "typical" usage would be something like e.g. `cut1 = h.cut_sqw(sqw_file, proj, l_limits=(l_min, l_max))`. `None` values are interpreted to use the full available range and standard step sizes.

Is that feasible, or am I missing some context here?

Reply:
@ajjackson The context is that it's the syntax used in Matlab which doesn't have key-word arguments. I agree it's an ugly syntax for Python, and your suggestion with default values would be better.

The range arguments are actually with respects to the `u`, `v`, `w` directions defined by the projection `proj` rather than `h`, `k`, `l` necessarily. It's only because I've defined in this case u||h, v||k, w||l that it's the case here (this projection could also be the default), giving:

```python
def cut_sqw(sqw_input: Union[Path, sqw],     # input is either path to sqw file or sqw object in memory
            proj: Dict[str, Any] = None,     # Defaults to: {'u':[1,0,0], 'v':[0,1,0], 'type':'rrr'}
            u_range: Optional[Tuple] = None,
            k_range: Optional[Tuple] = None,
            l_range: Optional[Tuple] = None,
            e_range: Optional[Tuple] = None) -> sqw    # output is sqw object in memory or file-backed
```